### PR TITLE
Complete smooth_l1_loss_op.

### LINKED ISSUE #3798: SmoothL1 Operator.
ref: [#1](https://github.com/rbgirshick/caffe-fast-rcnn/issues/1) and [#89](https://github.com/rbgirshick/py-faster-rcnn/issues/89)

This is a more general implementation than current paddle version.

### PR DESCRIPTION
**Pull request by pkuyym**
resolves #3798

### CODE REVIEW

File: paddle/operators/smooth_l1_loss_op.h [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License. */
+
+#pragma once
+#include "paddle/framework/eigen.h"
+#include "paddle/framework/op_registry.h"
```
Comment (QiJune) [inline paddle/operators/smooth_l1_loss_op.h]: Please include this header file:
```
#include "paddle/platform/hostdevice.h"
```

File: paddle/operators/smooth_l1_loss_op.h [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+template <typename T, int MajorType = Eigen::RowMajor,
+          typename IndexType = Eigen::DenseIndex>
+using EigenVector = framework::EigenVector<T, MajorType, IndexType>;
+template <typename T, int MajorType = Eigen::RowMajor,
+          typename IndexType = Eigen::DenseIndex>
+using EigenMatrix = framework::EigenMatrix<T, MajorType, IndexType>;
+
+template <typename T>
+struct SmoothL1LossFoward {
+  __host__ __device__ SmoothL1LossFoward(const T& sigma2) : sigma2(sigma2) {}
```
Comment (QiJune) [inline paddle/operators/smooth_l1_loss_op.h]: Please change 
```__host__ __device__```
 to 

```
HOSTDEVICE
```

File: paddle/operators/smooth_l1_loss_op.h [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+  __host__ __device__ T operator()(const T& val) const {
```
Comment (QiJune) [inline paddle/operators/smooth_l1_loss_op.h]: The same

File: paddle/operators/smooth_l1_loss_op.h [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        framework::make_ddim({static_cast<int>(in0->dims()[0]),
+                              static_cast<int>(in_counts / in0->dims()[0])});
+    auto errors_mat_view = EigenMatrix<T>::From(paddle_errors, mat_dims);
+    loss.device(place) = errors_mat_view.sum(Eigen::array<int, 1>({{1}}));
+  }
+};
+
+template <typename T>
+struct SmoothL1LossBackward {
+  __host__ __device__ SmoothL1LossBackward(const T& sigma2) : sigma2(sigma2) {}
```
Comment (QiJune) [inline paddle/operators/smooth_l1_loss_op.h]: The same

File: paddle/operators/smooth_l1_loss_op.h [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+  __host__ __device__ T operator()(const T& val) const {
```
Comment (QiJune) [inline paddle/operators/smooth_l1_loss_op.h]: The same

File: paddle/operators/smooth_l1_loss_op.h [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+using Tensor = framework::Tensor;
+template <typename T, int MajorType = Eigen::RowMajor,
+          typename IndexType = Eigen::DenseIndex>
+using EigenVector = framework::EigenVector<T, MajorType, IndexType>;
+template <typename T, int MajorType = Eigen::RowMajor,
+          typename IndexType = Eigen::DenseIndex>
+using EigenMatrix = framework::EigenMatrix<T, MajorType, IndexType>;
+
+template <typename T>
+struct SmoothL1LossFoward {
```
Comment (QiJune) [inline paddle/operators/smooth_l1_loss_op.h]: SmoothL1LossFoward -- > SmoothL1LossForward

File: paddle/operators/smooth_l1_loss_op.h [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+    auto mat_dims =
+        framework::make_ddim({static_cast<int>(in0->dims()[0]),
+                              static_cast<int>(in_counts / in0->dims()[0])});
+    auto errors_mat_view = EigenMatrix<T>::From(paddle_errors, mat_dims);
+    loss.device(place) = errors_mat_view.sum(Eigen::array<int, 1>({{1}}));
+  }
+};
+
+template <typename T>
+struct SmoothL1LossBackward {
```
Comment (QiJune) [inline paddle/operators/smooth_l1_loss_op.h]: Please unify the name Backward and Grad

File: paddle/operators/smooth_l1_loss_op.h (hunk 1 above)
Comment (pkuyym) [reply paddle/operators/smooth_l1_loss_op.h]: Done.

File: paddle/operators/smooth_l1_loss_op.h (hunk 2 above)
Comment (pkuyym) [reply paddle/operators/smooth_l1_loss_op.h]: Done.

File: paddle/operators/smooth_l1_loss_op.h (hunk 6 above)
Comment (pkuyym) [reply paddle/operators/smooth_l1_loss_op.h]: Done.

File: paddle/operators/smooth_l1_loss_op.h (hunk 3 above)
Comment (pkuyym) [reply paddle/operators/smooth_l1_loss_op.h]: Done.

File: paddle/operators/smooth_l1_loss_op.h (hunk 7 above)
Comment (pkuyym) [reply paddle/operators/smooth_l1_loss_op.h]: ```SmoothL1LossBackward``` is just a callback function used in ```SmoothL1LossGradKernel```

File: paddle/operators/smooth_l1_loss_op.h (hunk 4 above)
Comment (pkuyym) [reply paddle/operators/smooth_l1_loss_op.h]: Done.

File: paddle/operators/smooth_l1_loss_op.h (hunk 5 above)
Comment (pkuyym) [reply paddle/operators/smooth_l1_loss_op.h]: Done.

File: paddle/operators/smooth_l1_loss_op.h [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+        SmoothL1LossBackward<T>(sigma2));
+
+    auto* out0 = context.Output<Tensor>(framework::GradVarName("X"));
+    auto* out1 = context.Output<Tensor>(framework::GradVarName("Y"));
+
+    // compute weights
+    Tensor paddle_weights;
+    paddle_weights.mutable_data<T>(mat_dims, context.GetPlace());
+    auto weights = EigenMatrix<T>::From(paddle_weights);
+    // initialize to 1.0
```
Comment (QiJune) [inline paddle/operators/smooth_l1_loss_op.h]: Please remove 152-159, and change to:
```
weights.device(place) = weight.constant(static_cast<T>(1.0));
```

File: paddle/operators/smooth_l1_loss_op.h [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+
+    Tensor paddle_diff;
+    paddle_diff.mutable_data<T>({static_cast<int>(counts)}, context.GetPlace());
+    auto diff = EigenVector<T>::Flatten(paddle_diff);
+    // apply smooth l1 backwoard
+    diff.device(place) = EigenVector<T>::Flatten(*in2).unaryExpr(
+        SmoothL1LossBackward<T>(sigma2));
+
+    auto* out0 = context.Output<Tensor>(framework::GradVarName("X"));
+    auto* out1 = context.Output<Tensor>(framework::GradVarName("Y"));
```
Comment (QiJune) [inline paddle/operators/smooth_l1_loss_op.h]: This two lines code can move to line 172. It's better the place of declaration is close to the place of using.

File: paddle/operators/smooth_l1_loss_op.h (hunk 8 above)
Comment (pkuyym) [reply paddle/operators/smooth_l1_loss_op.h]: I tried weights.device(place) = weight.constant(static_cast<T>(1.0)); before, but it didn't work for GPU. Thanks for your suggestion, it works fine for both GPU and CPU.

File: paddle/operators/smooth_l1_loss_op.cc [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+                      framework::OpAttrChecker* op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("X", "Input of SmoothL1LossOp.");
+    AddInput("Y", "Target of SmoothL1LossOp.");
+    AddInput("InsideWeight", "Optional input to scale (X-Y).");
+    AddInput("OutsideWeight", "Optinal input to scale smooth l1 loss.");
+    AddOutput("diff", "Intermediate variable to cache Win*(X-Y).")
+        .AsIntermediate();
+    AddOutput("Out", "Final smooth l1 loss of inputs.");
+    AddComment(R"DOC(
```
Comment (Superjom) [inline paddle/operators/smooth_l1_loss_op.cc]: put AddComment in the end?

Comment (pkuyym) [reply paddle/operators/smooth_l1_loss_op.cc]: Done.

File: paddle/operators/smooth_l1_loss_op.cc [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+                            "Input of SmoothL1LossOp must be initialized.");
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("Y"),
+                            "Target of SmoothL1LossOp must be initialized.");
+
+    auto* x = ctx.Input<framework::Tensor>("X");
+    auto* y = ctx.Input<framework::Tensor>("Y");
+    PADDLE_ENFORCE_EQ(x->dims(), y->dims(),
+                      "Dimensions of SmoothL1LossOp's input and target "
+                      "must be same.");
+    PADDLE_ENFORCE_GE(framework::arity(x->dims()), 2,
```
Comment (qingqing01) [inline paddle/operators/smooth_l1_loss_op.cc]: `x->dims().size()`

File: paddle/operators/smooth_l1_loss_op.cc [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+                              "inside and outside weights.");
+      PADDLE_ENFORCE_EQ(inside_weight->dims(), x->dims(),
+                        "Dimensions of inside weight must be same with input.");
+      PADDLE_ENFORCE_EQ(
+          outside_weight->dims(), x->dims(),
+          "Dimensions of outside weight must be same with input.");
+    }
+
+    auto* diff = ctx.Output<framework::Tensor>("diff");
+    auto* out = ctx.Output<framework::Tensor>("Out");
```
Comment (qingqing01) [inline paddle/operators/smooth_l1_loss_op.cc]:
```cpp
auto* diff = ctx.Output<framework::LoDTensor>("diff");
auto* out = ctx.Output<framework::LoDTensor>("Out");
```

File: paddle/operators/smooth_l1_loss_op.cc [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+        .AsIntermediate();
+    AddOutput("Out", "Final smooth l1 loss of inputs.");
+    AddAttr<AttrType>("sigma", "Hyper parameter, default value is 3.0 .")
+        .SetDefault(3.0);
+    AddComment(R"DOC(
+Compute SmoothL1Loss for input and target.
+The equation is:
+loss = 0.5 * (sigma * (x - y)) ^ 2 if abs(x - y) < 1 / sigma^2
+       abs(x - y) - 0.5 / sigma^2  otherwise
+)DOC");
```
Comment (qingqing01) [inline paddle/operators/smooth_l1_loss_op.cc]: 77行后空一行，80行前空一行，公式缩进，有可能生成MarkDown，这样有缩进好些

File: paddle/operators/smooth_l1_loss_op.cc [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+
+ protected:
+  void InferShape(const framework::InferShapeContext& ctx) const override {
+    auto in_dims = ctx.Input<framework::Tensor>("X")->dims();
+    auto out_dims =
+        ctx.Input<framework::Tensor>(framework::GradVarName("Out"))->dims();
+    auto* x_grad = ctx.Output<framework::Tensor>(framework::GradVarName("X"));
+    auto* y_grad = ctx.Output<framework::Tensor>(framework::GradVarName("Y"));
+
+    PADDLE_ENFORCE_GE(framework::arity(out_dims), 2,
```
Comment (qingqing01) [inline paddle/operators/smooth_l1_loss_op.cc]: out_dims.size()

File: paddle/operators/smooth_l1_loss_op.cc [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+class SmoothL1LossOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  SmoothL1LossOpMaker(framework::OpProto* proto,
+                      framework::OpAttrChecker* op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("X", "Input of SmoothL1LossOp.");
+    AddInput("Y", "Target of SmoothL1LossOp.");
+    AddInput("InsideWeight", "Optional input to scale (X-Y).");
+    AddInput("OutsideWeight", "Optinal input to scale smooth l1 loss.");
+    AddOutput("diff", "Intermediate variable to cache Win*(X-Y).")
```
Comment (qingqing01) [inline paddle/operators/smooth_l1_loss_op.cc]: diff -> Diff ?

File: paddle/operators/smooth_l1_loss_op.h [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+
+    diff.device(place) = x - y;
+    // multiply inside weight
+    if (has_weight) {
+      auto inside_weight = EigenVector<T>::Flatten(*in2);
+      // cache diff, reused in bp
+      diff.device(place) = diff * inside_weight;
+    }
+
+    auto in_counts = framework::product(in0->dims());
```
Comment (qingqing01) [inline paddle/operators/smooth_l1_loss_op.h]:
```cpp
 auto in_counts = in0->numel();
```

use `tensor->numel()` to replace `framework::product`

File: paddle/operators/smooth_l1_loss_op.h [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+    Tensor paddle_errors;
+    paddle_errors.mutable_data<T>({static_cast<int>(in_counts)},
+                                  context.GetPlace());
+    auto errors = EigenVector<T>::Flatten(paddle_errors);
```
Comment (qingqing01) [inline paddle/operators/smooth_l1_loss_op.h]: Why the name called `paddle_errors` ?  :)

File: paddle/operators/smooth_l1_loss_op.h [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+
+    Tensor paddle_diff;
+    paddle_diff.mutable_data<T>({static_cast<int>(counts)}, context.GetPlace());
+    auto diff = EigenVector<T>::Flatten(paddle_diff);
+    // apply smooth l1 backwoard
+    diff.device(place) = EigenVector<T>::Flatten(*in2).unaryExpr(
+        SmoothL1LossBackward<T>(sigma2));
+
+    // compute weights
+    Tensor paddle_weights;
```
Comment (qingqing01) [inline paddle/operators/smooth_l1_loss_op.h]: 好奇为啥名字前面加上`paddle_` ?

File: python/paddle/v2/framework/tests/test_smooth_l1_loss_op.py [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+        sigma2 = sigma * sigma
+        diff = self.inputs['X'] - self.inputs['Y']
+        diff = diff * self.inputs['InsideWeight']
+        loss = np.vectorize(smooth_l1_loss_forward)(diff, sigma2)
+        loss = loss * self.inputs['OutsideWeight']
+        loss = loss.sum(1).reshape((dims[0], 1))
+        self.outputs = {'diff': diff, 'Out': loss}
+
+
+class SmoothL1LossGradOpTest(GradientChecker):
```
Comment (qingqing01) [inline python/paddle/v2/framework/tests/test_smooth_l1_loss_op.py]: Use the new unit testing framework.

File: paddle/operators/smooth_l1_loss_op.h [hunk 20]
```diff
@@ -0,0 +1,182 @@
...
+    auto* in2 = context.Input<Tensor>("InsideWeight");
+    auto* in3 = context.Input<Tensor>("OutsideWeight");
+    auto* out0 = context.Output<Tensor>("Diff");
+    auto* out1 = context.Output<Tensor>("Out");
+
+    out0->mutable_data<T>(context.GetPlace());
+    out1->mutable_data<T>(context.GetPlace());
+    auto place = context.GetEigenDevice<Place>();
+
+    auto sigma = static_cast<T>(context.op().Attr<AttrType>("sigma"));
```
Comment (qingqing01) [inline paddle/operators/smooth_l1_loss_op.h]: `context.op().Attr` -> `context.Attr`

File: python/paddle/v2/framework/tests/test_smooth_l1_loss_op.py [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+            ['Y'],
+            'Out',
+            max_relative_error=0.08,
+            no_grad_set=set(['X', 'InsideWeight', 'OutsideWeight']))
+
+    def test_check_grad_ingore_y(self):
+        self.check_grad(
+            ['X'],
+            'Out',
+            max_relative_error=0.08,
```
Comment (qingqing01) [inline python/paddle/v2/framework/tests/test_smooth_l1_loss_op.py]: tune max_relative_error smaller, better to smaller than 0.02.

Comment (pkuyym) [reply python/paddle/v2/framework/tests/test_smooth_l1_loss_op.py]: Done.

**Pull request merged**